src/sample.py

```diff
@@ -39,7 +40,7 @@ def message (self, buf, font, msg, data, _x_what_is_this):
 		return True
 debugger = Debugger()
 hb.buffer_set_message_func (buf, debugger.message, 1, 0)
-hb.buffer_add_utf8 (buf, text.encode('utf-8'), 0, -1)
+hb.buffer_add_codepoints (buf, codepoints, 0, -1)
```

Comment:
It accepts surrogate characters for one, which are invalid in UTF-32:
```python
a = "\uD800\uDFFF"
```
This is a valid string in Python, but not valid UTF-32 string and if you pass such code units to `_add_utf32()` it will replace them with the replacement character but `_add_codepoints()` won’t.